MaxText/layers/attentions.py

```diff
@@ -260,22 +241,174 @@ def key_rotary(self, key: Array, inputs_positions: Array):
       name='key_rotary')(inputs=key, position=inputs_positions)
     return key
 
-  def apply_attention(
+
+  # Following Pallas MHA Flash Attention Reference.
+  # https://github.com/google/jax/blob/main/jax/experimental/pallas/ops/tpu/flash_attention.py
+  def generate_attention_mask(
       self,
+      query,
+      key,
+      decoder_segment_ids: Array | None,
+      decode: bool = False,
+  ) -> Array | None:
+    mask = None
+    if decoder_segment_ids is not None:
+      mask = decoder_segment_ids[:, :, None] == decoder_segment_ids[:, None, :]
+      mask = mask[:, None, None,:, :]
+
+    _, q_seq_len, _, _ = query.shape
+    _, kv_seq_len, _, _ = key.shape
+    mask_shape = (q_seq_len, kv_seq_len)
+    row_ids = jax.lax.broadcasted_iota(jnp.int32, mask_shape, 0)
+    col_ids = jax.lax.broadcasted_iota(jnp.int32, mask_shape, 1)
+    causal_mask = (col_ids <= row_ids)[None, None, None, :, :]
+
+    mask = causal_mask if mask is None else jnp.logical_and(mask, causal_mask)
+    return jnp.where(mask, 0.0, DEFAULT_MASK_VALUE) if mask is not None else None
+  
+  def apply_attention(self,
       query: Array,
       key: Array,
       value: Array,
       decoder_segment_ids: Array | None,
-      attention_bias: Array | None,
       dropout_rng: PRNGKey | None,
       deterministic: bool,
-      decode: bool = False,
+      decode: bool = False) -> Array:
+    
+    if self.attention_kernel == "dot_product":
+      return self.apply_attention_dot(query, key, value, decoder_segment_ids, dropout_rng, deterministic, decode)
+    elif self.attention_kernel == 'flash':
+      if decode:
+        raise ValueError("""Decode not supported with flash attention.
+                            Use `dot_product` instead.""")
+      return self.tpu_flash_attention(query, key, value, decoder_segment_ids)
+    elif self.attention_kernel == 'gpu_flash' or self.attention_kernel == 'gpu_flash_triton':
+      if decode:
+        raise ValueError("""Decode not supported with flash attention.
+                            Use `dot_product` instead.""")
+      return self.gpu_flash_attention(query, key, value)
+    else:
+      raise ValueError('Unexpected attention type.')
+
+  def tpu_flash_attention(
+    self,
+    query: Array,
+    key: Array,
+    value: Array,
+    decoder_segment_ids: Array | None) -> Array:
+    """TPU Flash Attention."""
+    # Transpose to ('batch', 'heads', 'length', 'kv')
+    query = jax.numpy.transpose(query, axes=(0, 2, 1, 3))
```

Comment:
Any reason for not using `jnp.transpose` here?

Reply:
no, changed